Reintroduce generic "outputexports" target after CMake migration

As was expected, CMake migration broke building of 3rd-party projects with Zephyr ports which call out to Zephyr build system from their own (cross-platform) build systems.

Here's example: https://github.com/micropython/micropython/blob/master/ports/zephyr/Makefile#L70
~~~
$(MAKE) -f Makefile.zephyr BOARD=$(BOARD) CONF_FILE=$(CONF_FILE) initconfig outputexports lib
~~~

How this is supposed to work: The basic requirements are: 

- [x] being able to build 3rd-party app sources with the same compiler and compiler options as Zephyr uses; 
- [x] being able to include zephyr headers in the app sources (while building with app's build system, but with the compiler+options received per req. 1); 
- [x] being able to pass an application library for the final Zephyr link stage. 
- [ ] being able to do all of the above with the "default" CMake setup, without adding custom targets to every 3rd-party project CMakeLists.txt.

The process is thus:

1. Generate any Kconfig headers with config options, to make <zephyr.h>, etc. inclusion possible  - initconfig target.
2. Generate a special file Makefile.exports which contains dump of all make vars used for Zephyr builds (CC=... , CFLAGS=..., etc.) - outputexports
3. Being able to pass ALL_LIBS var to Zephyr's Makefile (and LDFLAGS_zephyr) to Zephyr "all" target (https://github.com/micropython/micropython/blob/master/ports/zephyr/Makefile.zephyr#L24)


It's unclear how to reinstate this functionality with CMake, please suggest.

@SebastianBoe :

> If one were to go down that road I might define a "custom_target" (see zephyr/cmake/usage/CMakeLists.txt). And have that custom_target call the zephyr_get_* API and write that information to a file.

Trying quick hack on that:

~~~
include($ENV{ZEPHYR_BASE}/cmake/extensions.cmake)

message("outputexports goes here")

zephyr_get_compile_options_as_string(cflags)

message("CFLAGS=${cflags}")
~~~

Leads to:

~~~
outputexports goes here
CMake Error at /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/cmake/extensions.cmake:158 (get_property):
  get_property could not find TARGET zephyr_interface.  Perhaps it has not
  yet been created.
Call Stack (most recent call first):
  /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/cmake/extensions.cmake:183 (zephyr_get_compile_options)
  /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/cmake/usage/outputexports.cmake:7 (zephyr_get_compile_options_as_string)
~~~

And indeed, I see references to zephyr_interface only in extensions.cmake, and nothing seems to "define" it there. (Per my understanding, there should be "add_<something>" call to define a target, a-la https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#binary-targets)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1497,6 +1497,11 @@ add_subdirectory(cmake/flash)
 add_subdirectory(cmake/usage)
 add_subdirectory(cmake/reports)
 
+add_subdirectory_ifdef(
+  CONFIG_MAKEFILE_EXPORTS
+  cmake/makefile_exports
+  )
+
 if(NOT CONFIG_TEST)
 if(CONFIG_ASSERT AND (NOT CONFIG_FORCE_NO_ASSERT))
   message(WARNING "__ASSERT() statements are globally ENABLED")

--- a/Kconfig.zephyr
+++ b/Kconfig.zephyr
@@ -332,6 +332,12 @@ config APPLICATION_DEFINED_SYSCALL
 	  Scan additional folders inside application source folder
 	  for application defined syscalls.
 
+config MAKEFILE_EXPORTS
+	bool "Generate build metadata files named Makefile.exports"
+	help
+	  Generates a file with build information that can be read by
+	  third party Makefile-based build systems.
+
 endmenu
 endmenu
 

--- a/cmake/makefile_exports/CMakeLists.txt
+++ b/cmake/makefile_exports/CMakeLists.txt
@@ -1,0 +1,29 @@
+# Copyright (c) 2020 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+set(exports
+  "
+CC = ${CMAKE_C_COMPILER}
+CXX = ${CMAKE_CXX_COMPILER}
+OBJCOPY = ${CMAKE_OBJCOPY}
+OBJDUMP = ${CMAKE_OBJDUMP}
+AS = ${CMAKE_AS}
+AR = ${CMAKE_AR}
+NM = ${CMAKE_NM}
+GDB = ${CMAKE_GDB}
+Z_CFLAGS = -I$<JOIN:$<TARGET_PROPERTY:zephyr_interface,INTERFACE_INCLUDE_DIRECTORIES>, -I> -isystem $<JOIN:$<TARGET_PROPERTY:zephyr_interface,INTERFACE_SYSTEM_INCLUDE_DIRECTORIES>, -isystem > -D$<JOIN:$<TARGET_PROPERTY:zephyr_interface,INTERFACE_COMPILE_DEFINITIONS>, -D> $<JOIN:$<TARGET_PROPERTY:zephyr_interface,INTERFACE_COMPILE_OPTIONS>, >
+"
+  )
+
+# file(GENERATE writes a file at Generation time. Also, it writes one
+# file per detected configuration, in this case, each COMPILE_LANGUAGE
+# is a new configuration.
+#
+# We use 'file(GENERATE' instead of configure_file because we want to
+# generate the file after Configure-time to have all the
+# metadata. Also, we don't use 'add_custom_command' because it cannot
+# read the generator expressions that we use.
+file(GENERATE
+  OUTPUT  ${CMAKE_BINARY_DIR}/Makefile.exports.$<COMPILE_LANGUAGE>
+  CONTENT "${exports}"
+)

